lupa/_lupa.pyx

```diff
@@ -1666,14 +1678,24 @@ cdef _LuaTable py_to_lua_table(LuaRuntime runtime, lua_State* L, items):
     check_lua_stack(L, 5)
     old_top = lua.lua_gettop(L)
     lua.lua_newtable(L)
+    cdef int lua_table_ref = lua.lua_gettop(L) # the index of the lua table which we are filling
     # FIXME: how to check for failure?
-
+    if recursive and mapped_objs is None:
+        mapped_objs = {}
     try:
         for obj in items:
+            if recursive:
+                if id(obj) not in mapped_objs:
+                    # this object is never seen before, we should cache it
+                    mapped_objs[id(obj)] = lua_table_ref
+                else:
+                    # this object has been cached, just get the corresponding lua table's index
+                    idx = mapped_objs[id(obj)]
+                    return new_lua_table(runtime, L, <int>idx)
```

Comment:
Ah, thanks for that comment, that actually helps. Then you're not using `mapped_objs` as it's named but rather as a `mapped_tables`, right? Seems worth renaming then to make the code clearer.